extensions/ros2/src/graph/Ros2PublishPointVelocityMarkersNode.cpp

```diff
@@ -0,0 +1,79 @@
+// Copyright 2023 Robotec.AI
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <graph/NodesRos2.hpp>
+#include <scene/Scene.hpp>
+
+void Ros2PublishPointVelocityMarkersNode::setParameters(const char* topicName, const char* frameId)
+{
+	this->frameId = frameId;
+	ros2InitGuard = Ros2InitGuard::acquire();
+	auto qos = rclcpp::QoS(10); // Use system default QoS
+	linesPublisher = ros2InitGuard->createUniquePublisher<visualization_msgs::msg::MarkerArray>(topicName, qos);
+}
+
+void Ros2PublishPointVelocityMarkersNode::validateImpl()
+{
+	IPointsNodeSingleInput::validateImpl();
+	if (!input->isDense()) {
+		throw InvalidPipeline(fmt::format("{} requires a compacted point cloud (dense)", getName()));
+	}
+}
+void Ros2PublishPointVelocityMarkersNode::enqueueExecImpl()
+{
+	pos->copyFrom(input->getFieldData(RGL_FIELD_XYZ_VEC3_F32));
+	vel->copyFrom(input->getFieldData(RGL_FIELD_ABSOLUTE_VELOCITY_VEC3_F32));
+
+	// Add special marker to delete previous markers
+	auto cancelMsg = visualization_msgs::msg::Marker();
+	cancelMsg.action = visualization_msgs::msg::Marker::DELETEALL;
+	cancelMsg.id = -1; // Avoid collision with lines id, RViz2 would complain
+
+	visualization_msgs::msg::MarkerArray lines;
+	lines.markers.emplace_back(cancelMsg);
+	lines.markers.emplace_back(makeLinesMarker());
+	if (!rclcpp::ok()) {
+		// TODO: This should be handled by the Graph.
+		throw std::runtime_error("Unable to publish a message because ROS2 has been shut down.");
+	}
+	linesPublisher->publish(lines);
+}
+
+const visualization_msgs::msg::Marker& Ros2PublishPointVelocityMarkersNode::makeLinesMarker()
+{
+	static visualization_msgs::msg::Marker marker;
```

Comment:
Making marker static we have hidden constrain - only one `Ros2PublishPointVelocityMarkersNode` allowed in the RGL. I'd make it a class member.

Reply:
Fixed